### PR TITLE
[9.0][mail_activity][mail_activity_calendar][mail_activity_crm]

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Implements the activities of 11.0 to 9.0 (and later 10.0), with some minor simplifications.

This PR includes the following modules:
* mail_activity
* mail_activity_calendar
* mail_activity_crm

Activities
==========

This module backports the activities from 11.0.

Known issues / Roadmap
======================

* The button 'Mark done' that appears for each activity in the chatter
  is not available.

* It is not possible to schedule a next activity based on a given activity

Mail Activity Calendar
======================

Adds the ability to create calendar meetings when you schedule an
activity of category 'Meeting'.

Known issues / Roadmap
======================

* When the user is redirected from the activity to the calendar view,
  the summary of the activity does not appear in the pop-up window following
  the creation of a new calendar event.

Activities in CRM
=================

This module adds activities to CRM.

Known issues / Roadmap
======================

* From the Kanban view you can only display the activities. You cannot
  schedule new activities from the kanban pop-up window, and the layout is
  slightly different.

* The Kanban view shows the next activity date and summary, in red if it
  is overdue.

### CODE REVIEW

File: mail_activity/views/mail_activity.xml
Comment (tarteo) [inline mail_activity/views/mail_activity.xml:77]: I've set 1 recommended next activity on a type:

image

When I select the activity type:

image

The radio button doesn't seem to appear.

File: mail_activity/static/src/less/systray.less
Comment (tarteo) [inline mail_activity/static/src/less/systray.less:4]: Can you improve the layout of the activitymenu items, please?
image

File: mail_activity/static/src/js/systray.js
Comment (tarteo) [inline mail_activity/static/src/js/systray.js:89]: You can also do `if(!data) { return; }` here. Makes the code flatter, but it's nitpicking so not really important :smile:

**Review by tarteo — CHANGES_REQUESTED** [review]
Thanks for the back port! I reviewed the module code and functionality and have some small comments.

File: mail_activity/static/src/less/systray.less
Comment (JordiBForgeFlow) [reply mail_activity/static/src/less/systray.less:4]: Looks better now :)

image

File: mail_activity/static/src/js/systray.js
Comment (JordiBForgeFlow) [reply mail_activity/static/src/js/systray.js:89]: I'd prefer the code as similar to the original as possible

File: mail_activity/views/mail_activity.xml
Comment (JordiBForgeFlow) [reply mail_activity/views/mail_activity.xml:77]: I have not fixed. But updated the README stating as known issue.

**Comment by JordiBForgeFlow** [comment]
@tarteo I have added mail_activity_calendar and mail_activity_crm. With the 3 modules the backport of activities to 9.0 and 10.0 will be much more complete!

**Comment by fuentes73** [comment]
Thanks, great job!  
Testing the module I've seen that if you schedule a call to a partner horizontal scroll is added to the view

captura de pantalla 2018-06-21 a las 10 21 03

A solution could be:
`.o_avatar_stack {position: relative}`

Plus, the action icon and it's border are set in white color. It should be in another color more visible,  isn't it?
captura de pantalla 2018-06-21 a las 10 31 57
 
captura de pantalla 2018-06-21 a las 10 41 59

**Comment by JordiBForgeFlow** [comment]
@fuentes010 I am not an expert in frontend development. Perhaps you know better and can make some quick adjustments? Please send me a PR with your proposed fixed, and hopefully we can fix this together.

**Review by tarteo — CHANGES_REQUESTED** [review]
@jbeficent Thanks! The calendar feature is really nice. I get an error though when I try to cancel the activity in the chatter.
`
Uncaught TypeError: self._rpc is not a function
http://3329956-286-b08677.runbot1.odoo-community.org/mail_activity_calendar/static/src/js/mail_activity.js:44
Traceback:
TypeError: self._rpc is not a function
    at Class.confirm_callback (http://3329956-286-b08677.runbot1.odoo-community.org/mail_activity_calendar/static/src/js/mail_activity.js:44:37)
    at HTMLButtonElement.<anonymous> (http://3329956-286-b08677.runbot1.odoo-community.org/web/static/src/js/framework/dialog.js:87:41)
    at HTMLButtonElement.dispatch (http://3329956-286-b08677.runbot1.odoo-community.org/web/static/lib/jquery/jquery.js:4641:58)
    at HTMLButtonElement.elemData.handle (http://3329956-286-b08677.runbot1.odoo-community.org/web/static/lib/jquery/jquery.js:4309:63)
`

I think instead of using `_rpc` you can use `rpc`. I think _rpc is introduced in 11.0 in the ServicesMixin in 9.0 the function is directly on web.Widget

**Comment by tarteo** [comment]
@jbeficent Can I create a PR on your fork / branch to fix the cancellation issue?

**Comment by JordiBForgeFlow** [comment]
Sure!

**Comment by fuentes73** [comment]
@jbeficent I've already done the [PR](https://github.com/Eficent/social/pull/1)

**Comment by JordiBForgeFlow** [comment]
Thanks Pablo!

**Comment by tarteo** [comment]
@jbeficent Made the fix https://github.com/Eficent/social/pull/2

**Review by fuentes73 — APPROVED** [review]

**Review by tarteo — APPROVED** [review]

**Review by hbrunn — DISMISSED** [review]
please fix https://travis-ci.org/OCA/social/jobs/396943416#L576

**Comment by JordiBForgeFlow** [comment]
@hbrunn We're good to go. Can you please merge?

**Comment by hbrunn** [comment]
can we have more test coverage?

**Comment by JordiBForgeFlow** [comment]
@hbrunn This would mean adding JS tests, which I am totally unable to add.. And adding python tests does not either make much sense, since the data model is pretty simple.

File: mail_activity/README.rst
Comment (odony) [inline mail_activity/README.rst:50]: Guys, do you have guidelines somewhere related to porting code from Odoo SA (or other third-party code)? The GPL licenses require that you identify specifically in each file what code is covered by what copyright and what license (when not uniform).

I find the current ports a little confusing, with a mix of licensing and copyright info. I see no copyright notice from Odoo SA (and its third-party contributors), and I can't tell by reading the source files who owns the copyright on what part of the code, nor what you changed (perhaps I simply missed it?)

For the record, the short notice "# Part of Odoo ..." only works inside the official Odoo code because:
 - it comes with matching LICENSE and COPYRIGHT files
 - external contributions are always subject to a CLA and authorship attribution is done via Git metadata on commits.
When you copy the code outside of official branches you can't replicate all of these conditions AFAICT, so you need to provide proper attribution, license and copyright info in each file, in context, and identify all your changes.

I know this stuff can be a PITA, but it's a key principle of open source, and I know we all want to do it right :-)

Comment (pedrobaeza) [reply mail_activity/README.rst:50]: @odony AFAIK, the module is full licensed as LGPL-3, which matches your license, isn't it?

Comment (odony) [reply mail_activity/README.rst:50]: @pedrobaeza well, there seems to be some conflicting info about that too, even in the manifest file ;-)
https://github.com/Eficent/social/blob/c3c1af5a6c02f80cad5158d3f6914311eebeb96b/mail_activity/__openerp__.py#L3

Comment (JordiBForgeFlow) [reply mail_activity/README.rst:50]: @odony I totally agree with you. I will add Odoo, S.A. wherever the code was moved, and change the licensing type. This feature is super useful, and unfortunately not all companies are able to move to 11.0 as quick as we'd like, so backporting is sometimes an option.

Comment (pedrobaeza) [reply mail_activity/README.rst:50]: You are right, @odony. Some copy/paste error, but @jbeficent is on it as he has said. Thanks for noticing.

**Comment by JordiBForgeFlow** [comment]
@odony I have added Odoo, S.A. explicitly in the copyrights section in all files. 
As for identifying the changes, the readme file in each module identifies, in the section 'Known issues / Roadmap' the differences with respect to 11.0. It is hard to identify differences explicitly in code. Code had do be adjusted quite a lot in the JS side, and in python code was extracted in bits and pieces from odoo 11.0, because in the core it was entangled with other  non activity-specific features.

**Comment by odony** [comment]
Thanks for updating the copyright mentions!


> As for identifying the changes, the readme file in each module identifies, in
> the section 'Known issues / Roadmap' the differences with respect to 11.0. It
> is hard to identify differences explicitly in code.

Sure, and if every file contains up-to-date mentions about the licenses and
copyright involved (or points to LICENSE/COPYRIGHT files that do), you have a
very defensible position that the license requirements are satisfied :)

Revision history can then be used to more precisely analyze ownership of code
lines. Especially if you have an initial commit to check in the original code,
and then build over that with separate commits (this could be part of some
guidelines)

**Comment by JordiBForgeFlow** [comment]
@odony In cases where Odoo discontinues a module, and someone wants to continue the work preserving commit history should be mandatory, and even simple using a standard code migration such as the ones OCA uses when migrating modules across major versions https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0. 


In this specific case this was not possible because we are not backporting an entire module, but specific features contained in various other modules. So preserving commit history seems impossible.

Thanks for your feedback anyway!

**Comment by rafaelbn** [comment]
Hi @hbrunn , i think you review is solved :smile: :+1:

**Comment by JordiBForgeFlow** [comment]
@hbrunn please notice 6e0429c. If you use some variation of mail_activity this is important to take into account.

**Comment by agb80** [comment]
Does this MR suppressed this other one: https://github.com/OCA/social/pull/286?

**Comment by paulius-sladkevicius** [comment]
Great initiative! Looking forward to test it out on v10, so if only @hbrunn is holding it to have merged and move to have it ready for v10?

**Comment by hbrunn** [comment]
oh, it's fivethousand years ago I reviewed this and I don't have time for proper review now, so I dismissed my review

**Comment by hbrunn** [comment]
PS: I'd still refuse to merge because it fails pylint currently...

**Comment by inspiredbusiness** [comment]
Hi,
Not sure if this was the best way to go about this, but we needed this in v10. There were issues merging this branch into v10 so that I could work on the port. I ended up copy/paste the files in this branch into a new v10 branch, committed, then did the work to port to v10. Let me know if there is anything I can do to get traction on this.

I have submitted a PR for this in v10

https://github.com/OCA/social/pull/328

**Comment by inspiredbusiness** [comment]
I have made changes in https://github.com/OCA/social/pull/328 that allows this to work with the crm activities report. Just had to add a hook to copy from mail compose to mail message and add the missing field to the crm report sql view. Like v12 the activity will not show in the activity report until it is completed and converted into a mail message

**Comment by LoisRForgeFlow** [comment]
@inspiredbusiness for v10 there was this PR: https://github.com/OCA/social/pull/204

**Comment by inspiredbusiness** [comment]
@lreficent I saw that PR but the first comment said it needed a UI to be useful. The branches were confusing and I could not get then to merge correctly and the v9 version seemed more complete. Feel free to ignore my PR. It seems all of the branches for this are stuck and I had a customer that could not wait any longer. So I ported v9 which seemed to be the most complete and have gone to production with the PR + a few other minor changes I never ended up pushing yet to GH. Let me know if there is anything I can do to make all of this get moving again.

**Comment by LoisRForgeFlow** [comment]
@inspiredbusiness Thank for the insights, from my side I just wanted to make you aware of the existence of another PR, maybe contributors of that other PR could give you more instructions on how to converge the work to a common point.

**Pull request closed without merging**
